File: installers/windows/tray/StatusReader.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TadaimaTray;

public class AgentStatus
{
    [JsonPropertyName("pid")] public int Pid { get; set; }
    [JsonPropertyName("version")] public string Version { get; set; } = "";
    [JsonPropertyName("connected")] public bool Connected { get; set; }
    [JsonPropertyName("relay")] public string Relay { get; set; } = "";
    [JsonPropertyName("deviceName")] public string DeviceName { get; set; } = "";
    [JsonPropertyName("activeDownloads")] public int ActiveDownloads { get; set; }
    [JsonPropertyName("lastHeartbeat")] public string LastHeartbeat { get; set; } = "";
    [JsonPropertyName("updateAvailable")] public string? UpdateAvailable { get; set; }
}

public class AgentConfig
{
    [JsonPropertyName("relay")] public string Relay { get; set; } = "";
    [JsonPropertyName("deviceToken")] public string DeviceToken { get; set; } = "";
    [JsonPropertyName("deviceId")] public string DeviceId { get; set; } = "";
    [JsonPropertyName("deviceName")] public string DeviceName { get; set; } = "";
    [JsonPropertyName("directories")] public DirectoriesConfig Directories { get; set; } = new();
    [JsonPropertyName("maxConcurrentDownloads")] public int MaxConcurrentDownloads { get; set; } = 2;
    [JsonPropertyName("rdPollInterval")] public int RdPollInterval { get; set; } = 30;

    public class DirectoriesConfig
    {
        [JsonPropertyName("movies")] public string Movies { get; set; } = "";
        [JsonPropertyName("tv")] public string Tv { get; set; } = "";
        [JsonPropertyName("staging")] public string Staging { get; set; } = "";
    }
}

public static class StatusReader
{
    private static readonly string ConfigDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "tadaima");

    public static string StatusPath => Path.Combine(ConfigDir, "status.json");
    public static string ConfigPath => Path.Combine(ConfigDir, "config.json");
    public static string LogPath => Path.Combine(ConfigDir, "logs", "tadaima.log");
    public static string AgentExePath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "Tadaima", "tadaima-agent.exe");

    public static AgentStatus? ReadStatus()
    {
        try
        {
            if (!File.Exists(StatusPath)) return null;
            var json = File.ReadAllText(StatusPath);
            return JsonSerializer.Deserialize<AgentStatus>(json);
        }
        catch { return null; }
    }

    public static AgentConfig? ReadConfig()
    {
        try
        {
            if (!File.Exists(ConfigPath)) return null;
            var json = File.ReadAllText(ConfigPath);
            return JsonSerializer.Deserialize<AgentConfig>(json);
        }
        catch { return null; }
    }

    public static void WriteConfig(AgentConfig config)
    {
        try
        {
            Directory.CreateDirectory(ConfigDir);
            var json = JsonSerializer.Serialize(config, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(ConfigPath, json);
        }
        catch { /* non-fatal */ }
    }

    public static bool IsAgentRunning()
    {
        var status = ReadStatus();
        if (status == null) return false;
        if (DateTime.TryParse(status.LastHeartbeat, out var heartbeat))
            return (DateTime.UtcNow - heartbeat.ToUniversalTime()).TotalSeconds < 30;
        return false;
    }
}

File: installers/macos/menubar/StatusReader.swift
import Foundation

/// Represents the agent's runtime status, read from ~/.config/tadaima/status.json
struct AgentStatus: Codable {
    let pid: Int
    let version: String
    let connected: Bool
    let relay: String
    let deviceName: String
    let activeDownloads: Int
    let lastHeartbeat: String
    let updateAvailable: String?
}

/// Represents the agent's config, read from ~/.config/tadaima/config.json
struct AgentConfig: Codable {
    var relay: String
    var deviceToken: String
    var deviceId: String
    var deviceName: String
    var directories: Directories
    var maxConcurrentDownloads: Int
    var rdPollInterval: Int

    struct Directories: Codable {
        var movies: String
        var tv: String
        var staging: String
    }
}

class StatusReader {
    static let shared = StatusReader()

    private let configDir: String
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init() {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        configDir = "\(home)/.config/tadaima"
        encoder.outputFormatting = .prettyPrinted
    }

    var statusPath: String { "\(configDir)/status.json" }
    var configPath: String { "\(configDir)/config.json" }
    var logPath: String { "\(configDir)/logs/tadaima.log" }

    func readStatus() -> AgentStatus? {
        guard let data = FileManager.default.contents(atPath: statusPath) else { return nil }
        return try? decoder.decode(AgentStatus.self, from: data)
    }

    func readConfig() -> AgentConfig? {
        guard let data = FileManager.default.contents(atPath: configPath) else { return nil }
        return try? decoder.decode(AgentConfig.self, from: data)
    }

    func writeConfig(_ config: AgentConfig) {
        guard let data = try? encoder.encode(config) else { return }
        FileManager.default.createFile(atPath: configPath, contents: data)
    }

    /// Returns true if the agent appears to be running (status file exists and heartbeat is recent)
    func isAgentRunning() -> Bool {
        guard let status = readStatus() else { return false }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let heartbeat = formatter.date(from: status.lastHeartbeat) else { return false }
        return Date().timeIntervalSince(heartbeat) < 30 // stale if >30s
    }
}
